Closes #4325: Add authenticated guest access feature to AZ CAS

joeparsons opened the pull request:
## Description
This PR adds a new "Guest Authentication Mode" feature to the AZ CAS module that allows sites to require CAS authentication for specific paths without creating Drupal user accounts. This addresses the issue of database bloat and management overhead caused by creating user accounts for every authenticated CAS user.

The feature works by:
1. Intercepting requests to configured paths that should require CAS authentication
2. Redirecting anonymous users to CAS login
3. Storing the authenticated CAS username in the session rather than creating a Drupal user account
4. Redirecting users back to the original path after authentication

The implementation includes:
• A new "Guest Authentication" section in the AZ CAS settings form
• Path-based configuration for specifying which paths should require CAS authentication
• Integration with the existing CAS module's authentication flow
• Support for existing Drupal users to continue logging in normally

Important Note on CAS Auto-Register Setting:
This feature requires the CAS module's "Auto register users" setting to be enabled, as it relies on the CAS authentication flow. To ensure proper functionality:
• When guest mode is enabled, the AZ CAS module automatically enables the CAS "Auto register users" setting
• The AZ CAS module adds a notice to the CAS settings form explaining this requirement
• The AZ CAS module prevents disabling the "Auto register users" setting while guest mode is active
• If a user attempts to disable "Auto register users" while guest mode is active, an error message is displayed and the setting is forced back to enabled

This ensures that the guest authentication feature works correctly while maintaining compatibility with the CAS module.

### Release notes
<!--- Delete this line (and the closing comment line) to enable release notes.

If this change requires release notes: provide a summary of changes, how to use this change, and any related links. This content will 
be pasted in the [release notes](https://github.com/az-digital/az_quickstart/releases). Use markdown format to ensure proper pasting of
information. [Release notes example](https://github.com/az-digital/az_quickstart/releases/tag/2.11.0)

Make sure to add the release notes label to this PR.

Add markdown of release notes here.


Delete this line to enable release notes.  -->


## Related issues
Closes #4325

_Note: Won't be possible to test this on ProboCI but it should be possible to test in a local development environment or on a Pantheon multidev._

## How to test
1. Enable Guest Mode:
   • Go to Administration → Configuration → AZ Quickstart → AZ CAS Settings
   • Check "Enable guest authentication mode"
   • Add paths to protect (e.g., /protected-content/*)
   • Save the configuration
   • Verify that the CAS module's "Auto register users" setting is automatically enabled

2. Create Test Content:
   • Create a page that matches your protected path pattern (e.g., /protected-content/test)
   • Create another page outside the protected path pattern

3. Test Anonymous Access:
   • Log out of Drupal
   • Try to access the unprotected page - should work normally
   • Try to access the protected page - should redirect to CAS login

4. Test Guest Authentication:
   • When redirected to CAS login, authenticate with valid NetID
   • You should be redirected back to the protected content
   • Check the user administration page - no new user account should be created

5. Test Existing User:
   • Log in as a regular Drupal user with CAS credentials
   • Access the protected page - should work without additional authentication
   • Your existing user account and permissions should function normally

6. Test CAS Settings Integration:
   • Go to Administration → Configuration → People → CAS
   • Try to disable the "Auto register users" setting
   • You should see an error message and the setting should remain enabled
   • Disable guest mode in AZ CAS settings
   • Now you should be able to disable the "Auto register users" setting

joeparsons committed:
WIP submodule.

joeparsons committed:
Move functionality into existing az_cas module, tie guest mode to CAS module forced_login settings.

joeparsons committed:
PHPCBF

joeparsons committed:
Working version.

joeparsons committed:
Settings form tweaks.

joeparsons commented:
Open questions:
- Would it be better to encapsulate this functionality in an optional submodule (`az_cas_guest`)?  I went back and forth on this a couple of times already and don't have a strong preference one way or the other at this point.
- Is the "list of paths" configuration schema provided in this PR sufficient?  Would it be better to implement implement something at the node level (a field or some other kind of node-specific setting) so that we could allow content editors / content admins to restrict access to certain nodes to authenticated NetID users?

joeparsons commented:
## Guest Mode Event Subscribers Overview

1. GuestAuthPathSubscriber
   • Listens to: KernelEvents::REQUEST (priority 35)
   • Purpose: Intercepts requests to paths configured for guest authentication
   • When an anonymous user visits a protected path:
     • Checks if the path matches any configured guest_auth_paths
     • If matched, redirects the user to CAS login with a destination parameter
   • This subscriber runs at a higher priority than the CAS module's forced login subscriber (priority 29) to ensure it processes first

2. CasPreRegisterSubscriber
   • Listens to: CasHelper::EVENT_PRE_REGISTER (priority 100)
   • Purpose: Prevents Drupal user account creation for guest users
   • After CAS authentication:
     • Checks if the user already has a Drupal account (if yes, allows normal login)
     • If not and guest mode is enabled, cancels automatic registration
     • Stores the CAS username in the session as a guest authentication
     • Gets the destination from the request query
     • Throws a GuestRedirectException with the destination URL

3. GuestRedirectExceptionSubscriber
   • Listens to: KernelEvents::EXCEPTION (priority 100)
   • Purpose: Handles redirection after guest authentication
   • Catches the GuestRedirectException thrown by CasPreRegisterSubscriber
   • Creates a TrustedRedirectResponse to the original destination
   • Sets this response on the event, completing the redirection

4. CasPreLoginSubscriber
   • Listens to: CasHelper::EVENT_PRE_LOGIN (priority 100)
   • Purpose: Prevents the CAS module from trying to log in a non-existent user
   • If a guest session is detected, cancels the login process
   • This prevents error messages about failed login attempts

## Authentication Flow

1. Anonymous user visits a path configured for guest authentication
2. GuestAuthPathSubscriber detects this and redirects to CAS login
3. User authenticates with CAS
4. CasPreRegisterSubscriber:
   • Prevents user account creation
   • Stores guest authentication in session
   • Throws GuestRedirectException
5. GuestRedirectExceptionSubscriber catches the exception and redirects to original path
6. On subsequent requests to protected paths:
   • GuestAuthPathSubscriber checks for guest authentication in session
   • If found, allows access without requiring re-authentication

This system effectively creates a "lightweight" authentication mechanism that verifies NetID credentials without creating Drupal user accounts, while maintaining compatibility with the existing CAS module and supporting regular user authentication for existing accounts.

trackleft commented on modules/custom/az_cas/az_cas.services.yml:
```diff
@@ -5,3 +5,27 @@ services:
+
+  az_cas.pre_register_subscriber:
+    class: Drupal\az_cas\EventSubscriber\CasPreRegisterSubscriber
+    arguments: ['@config.factory', '@logger.factory', '@request_stack', '@cas.user_manager', '@datetime.time']
```
```suggestion
    arguments: 
    -  @config.factory'
    - @logger.factory
    - @request_stack
    - @cas.user_manager'
    - @datetime.time
```

trackleft commented on modules/custom/az_cas/README.md:
```diff
@@ -1,15 +1,59 @@
-#### Packaged Dependencies
-- [CAS module](https://www.drupal.org/project/cas)
-- [External Authentication module](https://www.drupal.org/project/externalauth)
+- Drupal 9 or 10
+- CAS module
+- AZ Core module
```
```suggestion
#### Packaged Dependencies
- [CAS module](https://www.drupal.org/project/cas)
- [External Authentication module](https://www.drupal.org/project/externalauth)
```

trackleft commented on modules/custom/az_cas/src/Form/AzCasSettingsForm.php:
```diff
@@ -71,34 +72,89 @@ public function getFormId() {
+    $cas_settings_url = Url::fromRoute('cas.settings')->toString() . '#edit-user-accounts';
+    $form['guest_authentication']['guest_auth_settings']['auto_register_notice'] = [
+      '#type' => 'markup',
+      '#markup' => '<div class="messages messages--warning">' . $this->t('Guest authentication requires the CAS module\'s "Auto register users" setting to be enabled. This setting will be automatically enabled when guest mode is activated. <a href="@cas_settings_url">View CAS settings</a>', ['@cas_settings_url' => $cas_settings_url]) . '</div>',
```
We should add a reminder here that people should not be entering protected data even though this is "password protected".

trackleft replied:
We might also add some advice about the kinds of paths to add here, for example a canonical path like /node/33 could be considered better than a path alias like `/pages/photo-galleries`  because it is the permanent address unless deleted or unpublished of course.

trackleft commented on modules/custom/az_cas/config/install/az_cas.settings.yml:
```diff
@@ -1,3 +1,5 @@
 disable_admin_add_user_button: true
 disable_password_recovery_link: true
+guest_mode: false
+guest_auth_paths: []
```
Not sure if this is correct, but we can talk about it.
```suggestion
guest_auth_paths: {}
```

joeparsons commented on modules/custom/az_cas/az_cas.module:
```diff
@@ -20,6 +20,7 @@ function az_cas_help($route_name, RouteMatchInterface $route_match) {
       $output = '';
       $output .= '<h3>' . t('About') . '</h3>';
       $output .= '<p>' . t('Pre-configures contrib CAS module to work with University of Arizona WebAuth.') . '</p>';
+      $output .= '<p>' . t('The Quickstart CAS Guest Authentication functionality allows site visitors to authenticate via Quickstart CAS without creating Drupal user accounts.') . '</p>';
```
```suggestion
      $output .= '<p>' . t('The Quickstart CAS Guest Authentication functionality allows site visitors to authenticate via UA WebAuth without creating Drupal user accounts.') . '</p>';
```

joeparsons commented on modules/custom/az_cas/README.md:
```diff
@@ -1,15 +1,59 @@
+
+## Guest Authentication
+
+The guest authentication feature allows site visitors to authenticate via Quickstart CAS without creating individual Drupal user accounts. This is useful for sites that need to restrict content to authenticated NetID users but don't need to store user-specific data or provide personalized experiences.
```
```suggestion
The guest authentication feature allows site visitors to authenticate via University of Arizona WebAuth without creating individual Drupal user accounts. This is useful for sites that need to restrict content to authenticated NetID users but don't need to store user-specific data or provide personalized experiences.
```

joeparsons commented on modules/custom/az_cas/az_cas.services.yml:
```diff
@@ -5,3 +5,27 @@ services:
+
+  az_cas.guest_redirect_exception_subscriber:
+    class: Drupal\az_cas\EventSubscriber\GuestRedirectExceptionSubscriber
+    arguments: ['@logger.factory']
```
```suggestion
    arguments:
      - '@logger.factory'
```

joeparsons commented on modules/custom/az_cas/az_cas.services.yml:
```diff
@@ -5,3 +5,27 @@ services:
+
+  az_cas.guest_auth_path_subscriber:
+    class: Drupal\az_cas\EventSubscriber\GuestAuthPathSubscriber
+    arguments: ['@config.factory', '@path.matcher', '@current_user', '@logger.factory']
```
```suggestion
    arguments:
      - '@config.factory'
      - '@path.matcher'
      - '@current_user'
      - '@logger.factory'
```

joeparsons committed:
Apply suggestions from code review

Co-authored-by: Chris Green <chrisgreen@arizona.edu>

joeparsons committed:
Add session validation.

joeparsons commented:
Added session validation.

## Session Validation Enhancements

1. Created a GuestSessionManager Service:
   • Implemented a dedicated service class to handle all guest session operations
   • Added proper dependency injection for state, request stack, time, and configuration

2. Server-Side Session Storage:
   • Guest sessions are now stored in Drupal's state system
   • Each session includes:
     • CAS username
     • Creation timestamp
     • Expiration time (configurable)
     • IP address (for optional validation)

3. Comprehensive Session Validation:
   • Added multiple validation checks:
     • Verifies session exists in server-side storage
     • Checks session hasn't expired
     • Validates username matches
     • Optionally validates IP address matches

4. Configurable Security Options:
   • Added settings for session lifetime (default 8 hours)
   • Added option to enable/disable IP validation
   • These settings allow balancing security and usability

5. Automatic Session Cleanup:
   • Implemented cron job to remove expired sessions
   • Prevents state system bloat from abandoned sessions

6. Integration with Existing Flow:
   • Updated event subscribers to use the new validation system
   • Maintained compatibility with the CAS authentication flow

joeparsons committed:
Allow guest access for blocked CAS Drupal users.

joeparsons committed:
Add warning to settings form about regulated content.

joeparsons committed:
Reorder settings form notices.

joeparsons commented:
> I think we need a `hook_post_update_name` implemenation to add the two new config keys to the active configuration. https://www.specbee.com/blogs/update-and-post-update-hooks-for-successful-drupal-site-updates

I think Config Distro should cover this (?)

joeparsons committed:
Check aliases for system paths (and visa versa).  Add notice to node edit form for nodes with protected paths.

trackleft commented on modules/custom/az_cas/az_cas.module:
```diff
@@ -97,6 +107,84 @@ function az_cas_form_user_login_form_alter(&$form, FormStateInterface $form_stat
+ *
+ * Ensures auto_register remains enabled when guest mode is active.
+ */
+function az_cas_cas_settings_validate($form, FormStateInterface $form_state) {
```
```suggestion
function az_cas_settings_validate($form, FormStateInterface $form_state) {
```

joeparsons replied:
This validation function is specifically for a CAS module setting that gets updated by our form...

trackleft commented on modules/custom/az_cas/az_cas.module:
```diff
@@ -97,6 +107,84 @@ function az_cas_form_user_login_form_alter(&$form, FormStateInterface $form_stat
+    $form['user_accounts']['auto_register']['#disabled'] = TRUE;
+
+    // Add a custom validation to ensure auto_register remains enabled.
+    $form['#validate'][] = 'az_cas_cas_settings_validate';
```
```suggestion
    $form['#validate'][] = 'az_cas_settings_validate';
```

joeparsons committed:
Autoconfiguration for event subscribers is already enabled.  No service tags needed.

trackleft commented:
> > I think we need a `hook_post_update_name` implemenation to add the two new config keys to the active configuration. [specbee.com/blogs/update-and-post-update-hooks-for-successful-drupal-site-updates](https://www.specbee.com/blogs/update-and-post-update-hooks-for-successful-drupal-site-updates)
> 
> I think Config Distro should cover this (?)

Oh right, we have a distribution

joeparsons commented:
> Another idea before I forget. Put a message on nodes that are behind webauth with a link back to the webauth config form for disabling this if needed.
> 
> Alternatively add a field with a checkbox to nodes that allows putting them behind webauth.
> 
> This wouldn't work with views pages, or routes defined in code

Added a notice to the node edit form on nodes that are associated with guest auth paths.

tadean commented on modules/custom/az_cas/az_cas.module:
```diff
@@ -97,6 +107,84 @@ function az_cas_form_user_login_form_alter(&$form, FormStateInterface $form_stat
+/**
+ * Implements hook_form_alter().
+ */
+function az_cas_form_alter(&$form, FormStateInterface $form_state, $form_id) {
```
Can this use [hook_form_BASE_FORM_ID_alter()](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Form%21form.api.php/function/hook_form_BASE_FORM_ID_alter/11.x)?

tadean commented on modules/custom/az_cas/src/Service/GuestSessionManager.php:
```diff
@@ -0,0 +1,178 @@
+/**
+ * Service for managing guest authentication sessions.
+ */
+class GuestSessionManager {
```
Does this class potentially reinvent some functionality that's already available in the session service?

tadean commented on modules/custom/az_cas/src/Service/GuestSessionManager.php:
```diff
@@ -0,0 +1,178 @@
+
+    // Check IP address if configured to do so.
+    $config = $this->configFactory->get('az_cas.settings');
+    if ($config->get('guest_ip_validation') &&
```
I think there might be a lot of unexpected effects from this, particularly for load balanced proxies, roaming wireless access points, and dual IPv4 and IPv6 users.

trackleft committed:
Merge branch 'main' into issue/4325-cas-guest-access

joeparsons committed:
CAS 3.x deprecations

joeparsons committed:
Merge branch 'main' into issue/4325-cas-guest-access

joeparsons committed:
Merge branch 'main' into issue/4325-cas-guest-access
